Source/Scene/Model/ClassificationPipelineStage.js

```diff
@@ -0,0 +1,138 @@
+import defined from "../../Core/defined.js";
+import RuntimeError from "../../Core/RuntimeError.js";
+import ShaderDestination from "../../Renderer/ShaderDestination.js";
+import VertexAttributeSemantic from "../VertexAttributeSemantic.js";
+import ModelUtility from "./ModelUtility.js";
+
+/**
+ * The classification pipeline stage is responsible for batching features
+ * together to properly create draw commands in the
+ *
+ * @namespace ClassificationPipelineStage
+ *
+ * @private
+ */
+const ClassificationPipelineStage = {};
+ClassificationPipelineStage.name = "ClassificationPipelineStage"; // Helps with debugging
+
+/**
+ * Process a model. This modifies the following parts of the render resources:
+ *
+ * <ul>
+ *  <li>adds a define to the shader to indicate that the model classifies other assets</li>
+ *  <li>adds arrays containing batch lengths and offsets to the model's resources
+ * </ul>
+ *
+ * <p>
+ * See {@link ClassificationModelDrawCommand} for the use of the batch offsets / lengths.
+ * </p>
+ *
+ * @param {PrimitiveRenderResources} renderResources The render resources for this primitive.
+ * @param {ModelComponents.Primitive} primitive The primitive.
+ * @param {FrameState} frameState The frame state.
+ *
+ * @private
+ */
+ClassificationPipelineStage.process = function (
+  renderResources,
+  primitive,
+  frameState
+) {
+  const shaderBuilder = renderResources.shaderBuilder;
+
+  shaderBuilder.addDefine(
+    "HAS_CLASSIFICATION",
+    undefined,
+    ShaderDestination.BOTH
+  );
+
+  const runtimePrimitive = renderResources.runtimePrimitive;
+
+  if (!defined(runtimePrimitive.batchLengths)) {
+    const batchInfo = getClassificationBatchInfo(primitive);
+    const batchLengths = batchInfo.batchLengths;
+    const batchOffsets = batchInfo.batchOffsets;
+
+    runtimePrimitive.batchLengths = batchLengths;
+    runtimePrimitive.batchOffsets = batchOffsets;
```

Comment:
What if this function was `getClassificationBatchInfo(primitive, runtimePrimitive)` and modified `runtimePrimitive.batchLengths`/`runtimePrimitive.batchOffsets` in place? Then you wouldn't have to allocate a `batchInfo` object for the return value.